src/riak_kv_group_keys.erl

```diff
@@ -0,0 +1,547 @@
+        true ->
+            maybe_limit_accumulate(Entry, Bucket, GroupParams, FoldFun, Acc, Itr);
+        false ->
+            Acc
+    end;
+maybe_accumulate({{_DifferentBucket, _Key}, _BinaryValue}, _PrevEntry, _Bucket, _GroupParams, _FoldFun, Acc, _Itr) ->
+    Acc.
+
+%% this is a maybe_accumulate but we can't pattern match into Accumulator size
+maybe_limit_accumulate(Entry, Bucket, #group_params{max_keys = MaxKeys}=GroupParams, FoldFun, Acc, Itr) ->
+    case riak_kv_fold_buffer:size(Acc) =< MaxKeys of
+        true ->
+            accumulate(Entry, Bucket, GroupParams, FoldFun, Acc, Itr);
+        false -> Acc
+    end.
+
+%% Note that the two instances of `TargetBucket' below are intentional so that we accumulate only when
+%% we match the bucket that we're looking for.
+accumulate({{TargetBucket, Key}=BKey, BinaryValue},
+           TargetBucket,
+           GroupParams,
+           FoldFun,
+           Acc,
+           Itr) ->
+    NewAcc = try
+                 PrefixOrMeta = common_prefix_or_metadata(BKey, BinaryValue, GroupParams),
+                 FoldFun(TargetBucket, Key, PrefixOrMeta, Acc)
+             catch Error ->
+                     FoldFun(TargetBucket, Key, {error, Error}, Acc)
+             end,
+    enumerate(BKey, TargetBucket, GroupParams, Itr, FoldFun, NewAcc).
+
+common_prefix_or_metadata({_Bucket, Key} = BKey,
+                          BinaryValue,
+                          #group_params{prefix = Prefix, delimiter = Delimiter}) ->
+
+    case common_prefix(Key, Prefix, Delimiter) of
+        undefined -> {metadata, extract_metadata(BKey, BinaryValue)};
+        CommonPrefix -> {common_prefix, CommonPrefix}
+    end.
+
+common_prefix(_Key, _Prefix, undefined) ->
+    undefined;
+common_prefix(_Key, _Prefix, <<>>) ->
+    undefined;
+common_prefix(Key, undefined, Delimiter) ->
+    common_prefix(Key, <<>>, Delimiter);
+common_prefix(Key, Prefix, Delimiter) ->
+    Index = binary:longest_common_prefix([Key, Prefix]),
+    KeySansPrefix = binary_part(Key, Index, byte_size(Key) - Index),
+    Parts = binary:split(KeySansPrefix, Delimiter, [global]),
+    case length(Parts) > 1 andalso lists:all(fun(X) -> X =/= <<>> end, Parts) of
+        true ->
+            CommonPrefix = hd(Parts),
+            <<Prefix/binary, CommonPrefix/binary, Delimiter/binary>>;
+        false ->
+            undefined
+    end.
+
+extract_metadata({Bucket, Key} = _BKey,
+                 BinaryValue) ->
+    RObj = riak_object:from_binary(Bucket, Key, BinaryValue),
+    Contents = riak_object:get_contents(RObj),
+    case Contents of
+        [_Content] ->
+            riak_object:get_metadata(RObj);
+        _ ->
+            %% TODO
+            {error, riak_object_has_siblings}
+    end.
+
+%% 0xFF is the biggest byte for the purposes of sorting in LevelDB,
+%% used for seeking past all keys starting with the prefix.
+-spec append_max_byte(binary()) -> binary().
+append_max_byte(Binary) when is_binary(Binary) ->
```

Comment:
I think we decided that in this case it just breaks the optimization a bit of scanning past keys, so in the unlikely case a user writes a key like that we accept it. (it doesn't change the behavior of the consumer, just the efficiency of part of the scan)